# Исследование надежности заемщиков


## Введение 

Заказчик — кредитный отдел банка. 

Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. 

Входные данные от банка — статистика о платёжеспособности клиентов.


Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Описание данных:

`children` — количество детей в семье

`days_employed` — общий трудовой стаж в днях

`dob_years` — возраст клиента в годах

`education` — уровень образования клиента

`education_id` — идентификатор уровня образования

`family_status` — семейное положение

`family_status_id` — идентификатор семейного положения

`gender` — пол клиента

`income_type` — тип занятости

`debt` — имел ли задолженность по возврату кредитов

`total_income` — ежемесячный доход

`purpose` — цель получения кредита

## Загрузка данных и изучение общей информации

Импорт pandas и загрузка данных.

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

Вывод первых 20 строчек на экран.

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Выведем основную информацию о данных.

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

Выведем количество пропущенных значений для каждого столбца.

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце я обработаю на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

В данных есть аномалии, которые не отражают действительность и появились по какой-то ошибке. Например, отрицательное количество дней трудового стажа в столбце `days_employed`. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода `abs()`.

In [6]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится для исследования.

Выведем перечень уникальных значений столбца `children`.

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения.

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Убедимся, что все пропуски заполнены.

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв.

In [14]:
data['education'] = data['education'].str.lower()

Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. 

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Исследование данных

#### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [25]:
child_grouped = data.groupby('children').agg({'debt':['count','sum']})
child_grouped['conversion'] = round(child_grouped['debt']['sum'] / child_grouped['debt']['count'], 3)
child_grouped.head(6)


debt       conversion
          count   sum           
children                        
0         14091  1063      0.075
1          4808   444      0.092
2          2052   194      0.095
3           330    27      0.082
4            41     4      0.098
5             9     0      0.000

**Вывод:** Полученная таблица довольно показательная. По ней видно, что с увеличением количеством детей у заемщиков, понижается вероятность того, что они не выплатят долг. Более того, данная выборка говорит о том, что заемщики с 5 детьми всегда выплачивали кредит.

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [31]:
fam_id_grouped = data.groupby('family_status_id').agg({'debt':['count','sum']})
fam_id_grouped['conversion'] = round(fam_id_grouped['debt']['sum'] / fam_id_grouped['debt']['count'], 3)
fam_id_grouped.head(6)


debt      conversion
                  count  sum           
family_status_id                       
0                 12261  927      0.076
1                  4134  385      0.093
2                   951   63      0.066
3                  1189   84      0.071
4                  2796  273      0.098

**Вывод:** Таблица показывает, что в большинстве случаев наибольшая доля невыплачивающих кредит в срок среди заемщиков, которые либо не состоят в браке, либо состоят в гражданском браке. При этом по статистике наиболее добросовестные заемщики - вдовы или вдовцы.

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [34]:
inc_id_grouped = data.groupby('total_income_category').agg({'debt':['count','sum']})
inc_id_grouped['conversion'] = round(inc_id_grouped['debt']['sum'] / inc_id_grouped['debt']['count'], 3)
inc_id_grouped.head(6)

debt       conversion
                       count   sum           
total_income_category                        
A                         25     2      0.080
B                       5014   354      0.071
C                      15921  1353      0.085
D                        349    21      0.060
E                         22     2      0.091

**Вывод:** Зависимость между уровнем дохода заемщиков и возвратом кредита в срок существует. Анализ показывает, что заемщики с наименьшим уровнем дохода в большинстве случае являются должниками.

#### Как разные цели кредита влияют на его возврат в срок?

In [37]:
purpose_grouped = data.groupby('purpose_category').agg({'debt':['count','sum']})
purpose_grouped['conversion'] = round(purpose_grouped['debt']['sum'] / purpose_grouped['debt']['count'], 3)
purpose_grouped.head(6)

debt      conversion
                          count  sum           
purpose_category                               
операции с автомобилем     4279  400      0.093
операции с недвижимостью  10751  780      0.073
получение образования      3988  369      0.093
проведение свадьбы         2313  183      0.079

**Вывод:** Полученная таблица показывает, что наибольшая доля невыплативших в срок кредит среди заемщиков просивших финансирование на операции с автомобилем и получение образования.

#### Приведите возможные причины появления пропусков в исходных данных.

**Ответ:** Принимая во внимание то, что данные отсутствуют для столбцов days_employed и total_income, можно сделать вывод, что пропуски связаны с тем, что в данные столбцы вообще ничего не вводилось сотрудником банка. Так как заемщик не имеет опыта работы, и поэтому не имеет никакого дохода. В пользу этого говорит и то, что количство пропусков в этих столбцах одинаковое. Но это достаточно странная версия, потому как банк не должен выдавать кредит человеку без работы.

#### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

**Ответ:** Со статистической точки зрения медиана является самой надежной мерой центральной тенденции, не откликающейся на выбросы (аномалии) выборки.

## Заключение

Мною было проведено исследование о факторах, влияющих на платежеспособность заемщиков банка. Входными данными стали данные клиентов о их семейном статусе, уровне дохода, платежеспособности и др. Результаты исследования могут быть использованы при выработке банком новой кредитной политики в отношении граждан с разными материальными и социальными характеристиками.

Перед исследованием были выдвинуты следуюдующие гипотезы:
* Гипотеза 1: количество детей заемщиков влияет на погашение кредита в срок
* Гипотеза 2: семейный статус заемщиков влияет на погашение кредита в срок
* Гипотеза 3: уровень дохода заемщиков влияет на погащение кредита в срок
* Гипотеза 4: цель взятия кредия заемщиков влияет на погашение кредита в срок

Гипотеза 1 частично подтверждена:
- 9.8% заемщиков с 4 детьми являются задолжниками;
- 7.5% заемщиков без детей являются задолжниками;
- среди заемщиков с 5 детьми (всего 9 человек!) нет задолжников.

Выводы и рекомендации: На первый взгляд, заемщики без детей являются более платежеспособными, чем заемщики с детьми. Однако дополнительные статистические данные необходимы относительно заемщиков с 5 и более детьми.

Гипотеза 2 подтверждена:
- заемщики, не состоящие в браке или состоящие в гражданском браке, в большинстве случаев являются должниками (доля должников 9.8 и 9.3%, соответственно);
- овдовевшие заемщики более трепетно по сравнению с остальными относятся к финансовым обязательствам (доля должников 6.6%).

Выводы и рекомендации: так как доля должников между заемщиками с разным семейным статусом довольно невелика, то при выдаче им кредитов следует уделять внимание и другим их особенностям, таким как количество детей, уровень доходов и цели кредита.

Гипотеза 3 частично подтверждена:
- заемщики с минимальным уровнем дохода (категория Е) обладают большей долей должников (9.1%);
- однако заемщики с чуть более высоким уровнем дохода (категория D) обладают самой меньшей долей должников (6.0%).

Выводы и рекомендации: Следует настрожденно относиться к лицам с минимальным уровнем дохода, учитывать другие особенности заемщиков. Необходимо провести дополнительные исследования среди заемщиков с уровнем дохода категории D.

Гипотеза 4 подтверждена:
- по 9.3% заемщиков, запрашивающих кредит на операции с автомобилем и получение образования, являются должниками, что явлется наибольшим показателем;
- 7.3 и 7.9% доля должников среди заемщиков, запрашивающих финансирование на недвижимость и свадьбу, соответственно.

Выводы и рекомендаци: Так как разница между минимальной и максимальной долями должников в разных категориях довольно небольшая, то этому критерию следует уделять большее внимание лишь в сочетании с другими неблагополучными особенностями заемщиков.

**Общий вывод по исследованию:**

Исследование показало, что доли должников в разных социальных и материальных категориях заемщиков варьируют незначительно и отличаются лишь первыми процентами. Однако специалистам банков при предоставлении кредитов следует большее внимание уделять критериям, приведенным в гипотезах 2 и 4. Кроме того, предоставленная аналитикам выборка по некоторым категориям граждан обладает невысокой степенью репрезентативности. Необходим больший объем статистических данных.